libc/harness/test_exec.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libc-tests
+ *
+ * executable program for various exec functions
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <ctype.h>
+#include <unistd.h>
+#include <string.h>
+#include <libgen.h>
+
+extern char **environ;
+
+
+static void test_exec_execveEnv(bool changeEnv)
+{
+	char *const arg[] = { "/bin/to_exec", NULL };
+	if (changeEnv) {
+		char *const env[] = { "TEST1=exec_value", NULL };
+		setenv("TEST1", "invalid_value", 1);
+		setenv("TEST2", "should_dissapear", 1);
+
+		if (execve("/bin/test_exec", arg, env) == -1) {
+			fprintf(stderr, "execve function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+	else {
+		setenv("TEST1", "unchanged_value", 1);
+
+		if (execve("/bin/test_exec", arg, environ) == -1) {
+			fprintf(stderr, "execve function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+}
+
+
+static void test_exec_execvePath(void)
+{
+	char *const arg[] = { "to_exec", NULL };
+
+	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+
+	if (execve("test_exec", arg, environ) == -1) {
+		fprintf(stderr, "execve function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+static void test_exec_execvpeEnv(bool changeEnv)
+{
+	char *const arg[] = { "/bin/to_exec", NULL };
+	if (changeEnv) {
+		char *const env[] = { "TEST1=exec_value", NULL };
+		setenv("TEST1", "invalid_value", 1);
+		setenv("TEST2", "should_dissapear", 1);
+
+		if (execvpe("/bin/test_exec", arg, env) == -1) {
+			fprintf(stderr, "execvpe function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+	else {
+		setenv("TEST1", "unchanged_value", 1);
+
+		if (execvpe("/bin/test_exec", arg, environ) == -1) {
+			fprintf(stderr, "execvpe function failed\n");
+
+			exit(EXIT_FAILURE);
+		}
+	}
+}
+
+
+static void test_exec_execvpePath(void)
+{
+	char *const arg[] = { "to_exec", NULL };
+
+	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+
+	if (execvpe("test_exec", arg, environ) == -1) {
+		fprintf(stderr, "execve function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+static void test_exec_execvpEnv(void)
+{
+	char *const arg[] = { "/bin/to_exec", NULL };
+
+	setenv("TEST1", "unchanged_value", 1);
+
+	if (execvp("/bin/test_exec", arg) == -1) {
+		fprintf(stderr, "execvpe function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+static void test_exec_execvpPath(void)
+{
+	char *const arg[] = { "to_exec", NULL };
+
+	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin", 1);
+
+	if (execvp("test_exec", arg) == -1) {
+		fprintf(stderr, "execve function failed\n");
+
+		exit(EXIT_FAILURE);
+	}
+}
+
+
+int main(int argc, char *argv[], char const *envp[])
```

Comment:
it would probably be better to use `environ` variable instead of third parameter to `main()` as the first is POSIX-compliant and second is a GNU GCC extension (widely used but even so.. :)).